groups/etcd_group_service.go

```diff
@@ -0,0 +1,232 @@
+}
+
+// MemberGroups returns all groups which member takes part
+func (c *EtcdGroupService) MemberGroups(ctx context.Context, uid string) ([]string, error) {
+	prefix := memberGroupKey("", uid)
+	etcdRes, err := clientInstance.Get(ctx, prefix, clientv3.WithKeysOnly(), clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make([]string, etcdRes.Count)
+	for i, kv := range etcdRes.Kvs {
+		groups[i] = string(kv.Key)[len(prefix):]
+	}
+	return groups, nil
+}
+
+// Members returns all member's UID in current group
+func (c *EtcdGroupService) Members(ctx context.Context, groupName string) (map[string]*Payload, error) {
+	prefix := memberKey(groupName, "")
+	etcdRes, err := clientInstance.Get(ctx, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	members := make(map[string]*Payload, etcdRes.Count)
+	for _, kv := range etcdRes.Kvs {
+		payload := &Payload{}
+		if err = json.Unmarshal(kv.Value, payload); err != nil {
+			return nil, err
+		}
+		members[string(kv.Key)[len(prefix):]] = payload
+	}
+	return members, nil
+}
+
+// Member returns the Payload from User
+func (c *EtcdGroupService) Member(ctx context.Context, groupName, uid string) (*Payload, error) {
+	etcdRes, err := clientInstance.Get(ctx, memberKey(groupName, uid))
+	if err != nil {
+		return nil, err
+	}
+	payload := &Payload{}
+	if err = json.Unmarshal(etcdRes.Kvs[0].Value, payload); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+// Contains check whether a UID is contained in current group or not
+func (c *EtcdGroupService) Contains(ctx context.Context, groupName, uid string) (bool, error) {
+	etcdRes, err := clientInstance.Get(ctx, memberKey(groupName, uid), clientv3.WithCountOnly())
+	if err != nil {
+		return false, err
+	}
+	return etcdRes.Count > 0, nil
+}
+
+// Add adds UID and payload to group. If the group doesn't exist, it is created
+func (c *EtcdGroupService) Add(ctx context.Context, groupName, uid string, payload *Payload) error {
+	jsonPayload, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+	_, err = clientInstance.Put(ctx, memberKey(groupName, uid), string(jsonPayload), clientv3.WithLease(leaseID))
+	if err != nil {
+		return err
+	}
+	_, err = clientInstance.Put(ctx, memberGroupKey(groupName, uid), "", clientv3.WithLease(leaseID))
+	return err
+}
+
+// Leave removes specified UID from group
+func (c *EtcdGroupService) Leave(ctx context.Context, groupName, uid string) error {
+	_, err := clientInstance.Delete(ctx, memberKey(groupName, uid))
+	if err != nil {
+		return err
+	}
+	_, err = clientInstance.Delete(ctx, memberGroupKey(groupName, uid))
+	return err
+}
+
+// LeaveAll clears all UIDs in the group
+func (c *EtcdGroupService) LeaveAll(ctx context.Context, groupName string) error {
+	dResp, err := clientInstance.Delete(ctx, groupPrefix(groupName), clientv3.WithPrefix())
+	if err != nil {
+		return err
+	}
+	prefix := memberKey(groupName, "")
+	for _, kv := range dResp.PrevKvs {
```

Comment:
couldn't you use the Delete function with  `memberKey(groupName, "")` as prefix the same way you did in line 203?

Reply:
Nope, due to the fact that this will delete everything under pitaya/groups/{groupName}/... and this prefix is different from this one, which is: pitaya/groups/{groupName}/uids/...
This way I can pick the uid from the user deleted and also delete much more things, such as pitaya/groups/{groupName}/subgroups/...

Reply:
The way you are implementing this seems kind of hacky to me.
What happens if your sets desynchronize and you no longer have an user id in `groupPrefix(groupName)` but still needs to delete it from the other sets?

I'd delete all the necessary prefixes such as `pitaya/groups/{groupName}/uids/`, `pitaya/groups/{groupName}/subgroups/`, etc.

Reply:
With the command from line 203, it will delete everything related to this group. **pitaya/groups/{groupName}/uids/** will be deleted and so will **pitaya/groups/{groupName}/subgroups/** . The only set which is more problematic is the: **pitaya/uids/{uid}/groups/{groupName}** , which I need the uids to delete from... Do you think its a good idea to delete each one and get all the unique uids that I can find using your strategy?

Reply:
Ok, I understood the problem. My concern is that if `pitaya/groups/{groupName}/uids/` and `pitaya/uids/{uid}/groups/{groupName}` are out of sync for some reason we won't be able to resync them. I'm not sure what is the best way to ensure we won't have ghost groups in `pitaya/uids/{uid}/groups/{groupName} `.

Reply:
In the worst scenario, we have the leaseTTL, which will delete it in the long run. So, I don't think this will be a huge problem (or so I hope...)